File: tools/src/main/java/com/qeapp/tools/process/OperatedProcess.java
package com.qeapp.tools.process;

import org.jetbrains.annotations.NotNull;

public interface OperatedProcess extends Process {

    interface Observer extends Process.Observer {
        default void onProcessRunned(OperatedProcess process, double runnedProgress) {}
        default void onProcessStopped(OperatedProcess process, double stoppedProgress) {}
    }
    interface Observable extends FailableProcess {
        void addProcessObserver(@NotNull OperatedProcess.Observer observer);
        void removeProcessObserver(OperatedProcess.Observer observer);
    }

    /**
     * Instant end process
     */
    void end();

    /**
     * Pause process
     * @return current progress
     */
    double stop();

    /**
     * Continue process
     * @param progress Progress
     */
    void run(double progress);

    /**
     * Is stopped
     */
    boolean isStopped();

}

File: tools/src/main/java/com/kekadoc/tools/process/OperatedProcess.kt
package com.kekadoc.tools.process

interface OperatedProcess : Process {

    interface Observer : Process.Observer {
        fun onProcessRunned(process: OperatedProcess?, runnedProgress: Double) {}
        fun onProcessStopped(process: OperatedProcess?, stoppedProgress: Double) {}
    }
    interface Observable : FailableProcess {
        fun addProcessObserver(observer: Observer)
        fun removeProcessObserver(observer: Observer?)
    }

    /**
     * Instant end process
     */
    fun end()

    /**
     * Pause process
     * @return current progress
     */
    fun stop(): Double

    /**
     * Continue process
     * @param progress Progress
     */
    fun run(progress: Double)

    /**
     * Is stopped
     */
    val isStopped: Boolean

}